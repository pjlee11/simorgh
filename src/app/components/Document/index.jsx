import React from 'react';
import ResourceHints from './ResourceHints';
import '../../lib/globalStyles';
import { C_POSTBOX } from '../../lib/constants/styles';

/* eslint-disable react/prop-types */
const Document = ({ assets, app, data, styleTags, helmet }) => {
  const htmlAttrs = helmet.htmlAttributes.toComponent();
  const meta = helmet.meta.toComponent();
  const title = helmet.title.toComponent();
  const links = helmet.link.toComponent();
  const serialisedData = JSON.stringify(data);
  const scripts = assets.map(asset => (
    <script
      crossOrigin="anonymous"
      key={asset}
      type="text/javascript"
      src={asset}
      defer
    />
  ));

  return (
    <html lang="en-GB" {...htmlAttrs}>
      <head>
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet="utf-8" />
        <meta name="robots" content="noindex,nofollow" />
        <meta name="theme-color" content={C_POSTBOX} />
        {meta}
        <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />
        <link rel="manifest" href="/manifest.json" />
        <ResourceHints />
        {title}
        {links}
        {styleTags}
      </head>
      <body>
        {/* eslint-disable react/no-danger */
        /* disabling the rule that bans the use of dangerouslySetInnerHTML until a more appropriate implementation can be implemented */}
        <div id="root" dangerouslySetInnerHTML={{ __html: app }} />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.SIMORGH_DATA=${serialisedData}`,
          }}
        />
        {scripts}
      </body>
    </html>
  );
};

export default Document;
