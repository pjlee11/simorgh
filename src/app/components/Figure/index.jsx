import styled from 'styled-components';
import { GEL_SPACING_DBL } from '../../lib/constants/styles';

const Figure = styled.figure`
  margin: 0;
  padding-bottom: ${GEL_SPACING_DBL};
  width: 100%;
`;

export default Figure;
