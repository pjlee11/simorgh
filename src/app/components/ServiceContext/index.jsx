import React from 'react';
import { func, node, string } from 'prop-types';
import services from '../../lib/config/services';

/* Create ServiceContext using the default service */
export const ServiceContext = React.createContext(services.default);

export const ServiceContextProvider = ({ children, service }) => (
  <ServiceContext.Provider value={services[service]}>
    {children}
  </ServiceContext.Provider>
);

export const ServiceContextConsumer = ({ children }) => (
  <ServiceContext.Consumer>{children}</ServiceContext.Consumer>
);

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
};

ServiceContextConsumer.propTypes = {
  children: func.isRequired,
};
