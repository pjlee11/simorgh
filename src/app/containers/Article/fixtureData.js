import { blockContainingText, singleTextBlock } from '../../models/blocks';

const articleDataBuilder = (
  id,
  passportLanguage,
  home,
  headlineText,
  paragraphText,
  seoHeadline,
  promoHeadline,
  summary,
) => ({
  metadata: {
    id: `urn:bbc:ares::article:${id}`,
    locators: {
      optimoUrn: `urn:bbc:optimo:asset:${id}`,
    },
    type: 'article',
    createdBy: '',
    created: 1514808060000,
    firstPublished: 1514808060000,
    lastPublished: 1514811600000,
    lastUpdated: 1514811600000,
    passport: {
      language: passportLanguage,
      home,
      category: 'news',
      genre: null,
    },
    tags: {},
  },
  content: {
    model: {
      blocks: [
        blockContainingText('headline', headlineText),
        singleTextBlock(paragraphText),
      ],
    },
  },
  promo: {
    id: `urn:bbc:ares::article:${id}`,
    headlines: {
      seoHeadline,
      promoHeadline,
    },
    locators: {
      optimoUrn: `urn:bbc:optimo:asset:${id}`,
    },
    summary,
    timestamp: 1514811600000,
  },
});

export const articleDataNews = articleDataBuilder(
  'c0000000001o',
  'en-gb',
  'http://www.bbc.co.uk/ontologies/passport/home/News',
  'Article Headline',
  'A paragraph.',
  'Article Headline for SEO',
  'Article Headline for Promo',
  'Article summary.',
);

export const articleDataPersian = articleDataBuilder(
  'cwv2xv848j5o',
  'fa',
  'http://www.bbc.co.uk/ontologies/passport/home/Persian',
  'سرصفحه مقاله',
  'یک پاراگراف.',
  'سرصفحه مقاله',
  'سرصفحه مقاله برای ارتقاء',
  'خلاصه مقاله',
);
