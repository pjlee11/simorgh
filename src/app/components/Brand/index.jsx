import React from 'react';
import styled from 'styled-components';
import { bool } from 'prop-types';
import { ServiceContextConsumer } from '../ServiceContext';
import VisuallyHiddenText from '../VisuallyHiddenText';
import {
  C_POSTBOX,
  C_WHITE,
  GEL_SPACING,
  GEL_SPACING_DBL,
  GEL_SPACING_HLF,
  group3ScreenWidthMax,
} from '../../lib/constants/styles';
import {
  layoutGridWrapper,
  layoutGridItemConstrained,
  layoutGridItem,
} from '../../lib/layoutGrid';

const SVG_TOP_OFFSET = '1.25rem'; // 20px
const SVG_BOTTOM_OFFSET = '1.5rem'; // 24px
const BANNER_HEIGHT = '5rem'; // 80px

const SVG_HEIGHT_PX = 24;
const SVG_HEIGHT = `${SVG_HEIGHT_PX / 16}rem`;
const SVG_WIDTH_PX = 167.95;
const SVG_WIDTH = `${SVG_WIDTH_PX / 16}rem`;

const VIEWBOX_VALUES = `0 0 ${SVG_WIDTH_PX} ${SVG_HEIGHT_PX}`;

const StyledGridWrapper = styled.div`
  ${layoutGridWrapper};
  background-color: ${C_POSTBOX};
  height: ${BANNER_HEIGHT};
  width: 100%;
  @media (min-width: ${group3ScreenWidthMax}) {
    padding: 0 ${GEL_SPACING_DBL};
  }
`;

const StyledLink = styled.a`
  padding-top: ${GEL_SPACING};
  display: inline-block;
`;

const StyledSpan = styled.span`
  display: block;
  padding-bottom: ${SVG_BOTTOM_OFFSET};
  ${StyledLink}:hover &,
  ${StyledLink}:focus & {
    text-decoration: none;
    border-bottom: ${GEL_SPACING_HLF} solid ${C_WHITE};
  }
`;

const BrandSvg = styled.svg`
  display: block;
  height: ${SVG_HEIGHT};
  width: ${SVG_WIDTH};
  margin-top: ${SVG_TOP_OFFSET};
`;

const Brand = ({ indentedLogo }) => {
  const StyledGridItem = styled.div`
    ${indentedLogo ? layoutGridItemConstrained : layoutGridItem};
  `;

  return (
    <StyledGridWrapper>
      <StyledGridItem indentedLogo={indentedLogo}>
        <StyledLink href="https://www.bbc.co.uk/news">
          <StyledSpan>
            <BrandSvg
              viewBox={VIEWBOX_VALUES}
              xmlns="http://www.w3.org/2000/svg"
              focusable="false"
              aria-hidden="true"
            >
              <g fill="#fff" fillRule="evenodd">
                <path d="M84.32,0V24H58.82V0ZM79.26,16l-.41.25a11.91,11.91,0,0,1-6.07,1.85c-4.18,0-6.93-2.49-6.94-6.09s2.88-6.13,6.83-6.14a12,12,0,0,1,6,1.71l.4.22V4.6l-.17-.07A16.54,16.54,0,0,0,72.7,3.14a10.56,10.56,0,0,0-7.19,2.57,8.68,8.68,0,0,0-2.85,6.54A8.46,8.46,0,0,0,65,17.93a10,10,0,0,0,7.5,2.94h0a14,14,0,0,0,6.56-1.5l.15-.07Z M54.94,0V24H29.43V0ZM48.88,15.91a4.62,4.62,0,0,0-3.77-4.47,4.25,4.25,0,0,0,1.49-1.19,3.75,3.75,0,0,0,.72-2.34,4.17,4.17,0,0,0-1.39-3.13,6.33,6.33,0,0,0-4.36-1.41H36.12V20.65H42.5a7.18,7.18,0,0,0,4.9-1.53A4.27,4.27,0,0,0,48.88,15.91Z M45.7,15.57A2,2,0,0,1,45,17.16a4.22,4.22,0,0,1-2.85.79H39.22V13.23H42a4.89,4.89,0,0,1,2.82.68A1.93,1.93,0,0,1,45.7,15.57Z M43.32,9.92a2,2,0,0,0,.82-1.79,1.82,1.82,0,0,0-.58-1.43,3.33,3.33,0,0,0-2.26-.63H39.22v4.47h1.43A4.62,4.62,0,0,0,43.32,9.92Z M25.55,0V24H0V0ZM19.49,15.91a4.62,4.62,0,0,0-3.77-4.47,4.24,4.24,0,0,0,1.49-1.19,3.75,3.75,0,0,0,.72-2.34,4.18,4.18,0,0,0-1.39-3.13,6.33,6.33,0,0,0-4.36-1.41H6.74V20.65h6.38A7.18,7.18,0,0,0,18,19.12,4.27,4.27,0,0,0,19.49,15.91Z M13.93,9.92a2,2,0,0,0,.82-1.79,1.82,1.82,0,0,0-.58-1.43,3.33,3.33,0,0,0-2.26-.63H9.83v4.47h1.43A4.62,4.62,0,0,0,13.93,9.92Z M15.47,13.91a4.89,4.89,0,0,0-2.82-.68H9.83V18h2.94a4.23,4.23,0,0,0,2.85-.79,2,2,0,0,0,.69-1.59A1.93,1.93,0,0,0,15.47,13.91Z" />
                <path d="M111,3.36h2.36V20.7h-2.13L99.66,7.35V20.7H97.33V3.36h2L111,16.83Z" />
                <path d="m117.86 3.36h9.83v2.21h-7.35v5.29h7.1v2.22h-7.1v5.39h7.58v2.21h-10.06z M154.18,3.36h2.48l-7,17.41h-.55l-5.67-14.1-5.73,14.1h-.53l-7-17.41h2.5l4.78,12,4.81-12h2.35l4.83,12Z M163.38,13.43l-1.89-1.15A8.57,8.57,0,0,1,159,10.16a4,4,0,0,1-.75-2.41,4.26,4.26,0,0,1,1.42-3.33,5.31,5.31,0,0,1,3.69-1.28,7,7,0,0,1,4,1.22V7.17a5.74,5.74,0,0,0-4-1.8,3.34,3.34,0,0,0-2,.56,1.71,1.71,0,0,0-.78,1.44,2.22,2.22,0,0,0,.58,1.46,7.25,7.25,0,0,0,1.85,1.43l1.9,1.12Q168,13.28,168,16.21a4.42,4.42,0,0,1-1.4,3.39A5.11,5.11,0,0,1,163,20.9a7.63,7.63,0,0,1-4.68-1.58V16.17a5.84,5.84,0,0,0,4.65,2.55,2.93,2.93,0,0,0,1.94-.65,2,2,0,0,0,.78-1.63Q165.67,14.85,163.38,13.43Z" />
              </g>
            </BrandSvg>
            <ServiceContextConsumer>
              {({ brandName }) => (
                <VisuallyHiddenText>{brandName}</VisuallyHiddenText>
              )}
            </ServiceContextConsumer>
          </StyledSpan>
        </StyledLink>
      </StyledGridItem>
    </StyledGridWrapper>
  );
};

Brand.defaultProps = {
  indentedLogo: true,
};

Brand.propTypes = {
  indentedLogo: bool,
};

export default Brand;
