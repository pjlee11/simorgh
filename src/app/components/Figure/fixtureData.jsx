import React from 'react';
import FigureContainer from '../../containers/Figure';
import { ServiceContext } from '../ServiceContext';

const imageAlt = 'Pauline Clayton';
const imageSrc =
  'https://ichef.bbci.co.uk/news/640/cpsprodpb/E7DB/production/_101655395_paulineclayton.jpg';
const imageRatio = 56.25;
const captionValue = 'This is a caption';
const captionValueContainingLink =
  'This is a caption [with a link](https://bbc.com/news)';
const copyrightText = 'Getty Images';

const serviceContextStubNews = {
  imageCaptionOffscreenText: 'Image caption, ',
};

const generateFixtureData = (caption, copyright) => (
  <ServiceContext.Provider value={serviceContextStubNews}>
    <FigureContainer
      src={imageSrc}
      alt={imageAlt}
      ratio={imageRatio}
      caption={caption}
      copyright={copyright}
    />
  </ServiceContext.Provider>
);

export const FigureImage = generateFixtureData();

export const FigureImageWithCaption = generateFixtureData(captionValue);

export const FigureImageWithCaptionContainingLink = generateFixtureData(
  captionValueContainingLink,
);

export const FigureImageWithCopyright = generateFixtureData(
  null,
  copyrightText,
);

export const FigureImageWithCopyrightAndCaption = generateFixtureData(
  captionValue,
  copyrightText,
);
