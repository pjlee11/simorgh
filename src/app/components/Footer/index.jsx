import React from 'react';
import styled from 'styled-components';
import { arrayOf, shape, string } from 'prop-types';
import Brand from '../Brand';
import Link from './Link';
import FooterList from './List';
import {
  C_ORBIT_GREY,
  C_WHITE,
  FF_NEWS_SANS_REG,
  GEL_SPACING_DBL,
} from '../../lib/constants/styles';
import { layoutGridWrapper, layoutGridItem } from '../../lib/layoutGrid';
import { T_BREVIER } from '../../lib/constants/typography';

const StyledFooterLinksWrapper = styled.div`
  ${layoutGridWrapper};
  background-color: ${C_ORBIT_GREY};
`;

const StyledFooterLinks = styled.div`
  ${layoutGridItem}
  ${T_BREVIER}
  font-family: ${FF_NEWS_SANS_REG};
`;

const StyledParagraph = styled.p`
  color: ${C_WHITE};
  margin: 0;
  padding: ${GEL_SPACING_DBL} 0;
`;

const Footer = ({ links, copyrightText, externalLink }) => (
  <footer role="contentinfo">
    <Brand indentedLogo={false} />
    <StyledFooterLinksWrapper>
      <StyledFooterLinks>
        <FooterList links={links} />
        <StyledParagraph>
          {copyrightText}
          <Link text={externalLink.text} href={externalLink.href} inline />
        </StyledParagraph>
      </StyledFooterLinks>
    </StyledFooterLinksWrapper>
  </footer>
);

const linkPropTypes = shape({
  href: string.isRequired,
  text: string.isRequired,
});

Footer.propTypes = {
  links: arrayOf(linkPropTypes.isRequired).isRequired,
  copyrightText: string.isRequired,
  externalLink: linkPropTypes.isRequired,
};

export default Footer;
