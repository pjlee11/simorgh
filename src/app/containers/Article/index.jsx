import React, { Fragment } from 'react';
import { bool, string, shape } from 'prop-types';
import styled from 'styled-components';
import MetadataContainer from '../Metadata';
import Header from '../../components/Header';
import Footer from '../Footer';
import headings from '../Headings';
import text from '../Text';
import image from '../Image';
import Blocks from '../Blocks';
import articlePropTypes from '../../models/propTypes/article';
import { ServiceContextProvider } from '../../components/ServiceContext';
import Timestamp from '../../components/Timestamp';
import {
  layoutGridWrapper,
  layoutGridItemConstrained,
} from '../../lib/layoutGrid';
import { C_OAT_LHT } from '../../lib/constants/styles';

const Wrapper = styled.div`
  ${layoutGridWrapper};
`;

const OatWrapper = styled(Wrapper)`
  background: ${C_OAT_LHT};
`;

const GridItemConstrained = styled.div`
  ${layoutGridItemConstrained};
`;

const componentsToRenderHeadline = {
  headline: headings,
};

const componentsToRenderMain = {
  subheadline: headings,
  text,
  image,
};

const splitBlocksByHeadline = ({ model }) => {
  const { blocks } = model;

  const headlineIndexPlusOne =
    blocks.findIndex(({ type }) => type === 'headline') + 1;

  const headlineBlocks = blocks.slice(0, headlineIndexPlusOne);
  const mainBlocks = blocks.slice(headlineIndexPlusOne, blocks.length);

  return { headlineBlocks, mainBlocks };
};

/*
  [1] This handles async data fetching, and a 'loading state', which we should look to handle more intelligently.
*/
const ArticleContainer = ({ loading, error, data }) => {
  if (loading) return 'Loading...'; /* [1] */
  if (error) return 'Something went wrong :(';
  if (data) {
    const { isAmp, data: articleData, service } = data;
    const { content, metadata, promo } = articleData;

    const { headlineBlocks, mainBlocks } = splitBlocksByHeadline(content);

    /*
    * headlineBlocks length check is temporary
    * Simorgh will respond with 400 to lack of headline block in issue 
    * https://github.com/BBC-News/simorgh/issues/836
    */
    if (headlineBlocks.length > 0) {
      return (
        <Fragment>
          <ServiceContextProvider service={service}>
            <Header />
            <MetadataContainer
              isAmp={isAmp}
              metadata={metadata}
              promo={promo}
              service={service}
            />
            <main role="main">
              <Wrapper>
                <GridItemConstrained>
                  <Blocks
                    blocks={headlineBlocks}
                    componentsToRender={componentsToRenderHeadline}
                  />
                  <Timestamp timestamp={metadata.lastUpdated} />
                </GridItemConstrained>
              </Wrapper>
              <OatWrapper>
                <GridItemConstrained>
                  <Blocks
                    blocks={mainBlocks}
                    componentsToRender={componentsToRenderMain}
                  />
                </GridItemConstrained>
              </OatWrapper>
            </main>
            <Footer />
          </ServiceContextProvider>
        </Fragment>
      );
    }
  }

  return null;
};

ArticleContainer.propTypes = {
  loading: bool,
  error: string,
  data: shape(articlePropTypes),
};

ArticleContainer.defaultProps = {
  loading: false,
  error: null,
  data: null,
};

export default ArticleContainer;
