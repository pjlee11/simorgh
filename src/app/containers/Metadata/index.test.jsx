import React from 'react';
import renderer from 'react-test-renderer';
import Helmet from 'react-helmet';
import MetadataContainer from './index';
import { ServiceContextProvider } from '../../components/ServiceContext';
import { articleDataNews, articleDataPersian } from '../Article/fixtureData';
import { shouldShallowMatchSnapshot } from '../../helpers/tests/testHelpers';
import services from '../../lib/config/services/index';

const MetadataWithContextAsObject = (service, serviceFixtureData) => {
  const { metadata, promo } = serviceFixtureData;

  renderer.create(
    <ServiceContextProvider service={service}>
      <MetadataContainer
        isAmp={false}
        metadata={metadata}
        promo={promo}
        service={service}
      />
    </ServiceContextProvider>,
  );

  return Helmet.peek();
};

describe('no data', () => {
  shouldShallowMatchSnapshot(
    'should render null',
    <MetadataContainer isAmp={false} metadata={{}} promo={{}} service="" />,
  );
});

const articleMetadataBuilder = (
  serviceName,
  lang,
  title,
  description,
  id,
  seoTitle,
) => {
  const serviceConfig = services[serviceName];

  return {
    htmlAttributes: { lang },
    linkTags: [
      {
        rel: 'canonical',
        href: `https://www.bbc.com/${serviceConfig.service}/articles/${id}`,
      },
    ],
    metaTags: [
      {
        name: 'viewport',
        content: 'width=device-width, initial-scale=1, minimum-scale=1',
      },
      {
        name: 'article:author',
        content: serviceConfig.articleAuthor,
      },
      {
        name: 'article:modified_time',
        content: '2018-01-01T13:00:00.000Z',
      },
      {
        name: 'article:published_time',
        content: '2018-01-01T12:01:00.000Z',
      },
      { name: 'description', content: description },
      { name: 'fb:admins', content: 100004154058350 },
      { name: 'fb:app_id', content: 1609039196070050 },
      { name: 'og:description', content: description },
      {
        name: 'og:image',
        content: serviceConfig.defaultImage,
      },
      { name: 'og:image:alt', content: serviceConfig.brandName },
      { name: 'og:locale', content: serviceConfig.locale },
      { name: 'og:site_name', content: serviceConfig.brandName },
      { name: 'og:title', content: seoTitle },
      { name: 'og:type', content: 'article' },
      {
        name: 'og:url',
        content: `https://www.bbc.com/${serviceConfig.service}/articles/${id}`,
      },
      { name: 'twitter:card', content: 'summary_large_image' },
      { name: 'twitter:creator', content: serviceConfig.twitterCreator },
      { name: 'twitter:description', content: description },
      { name: 'twitter:image:alt', content: serviceConfig.brandName },
      {
        name: 'twitter:image:src',
        content: serviceConfig.defaultImage,
      },
      { name: 'twitter:site', content: serviceConfig.twitterSite },
      { name: 'twitter:title', content: seoTitle },
    ],
    title: [seoTitle, ' – ', serviceConfig.brandName],
  };
};

const doesMatch = (result, fixture) => {
  Object.keys(fixture).forEach(key => {
    expect(result[key]).toEqual(fixture[key]);
  });
};

describe('Successfully passes data to the Metadata component via React context', () => {
  it('it should pass news data to the Metadata component', () => {
    doesMatch(
      MetadataWithContextAsObject('news', articleDataNews),
      articleMetadataBuilder(
        'news',
        'en-gb',
        'Article Headline',
        'Article summary.',
        'c0000000001o',
        'Article Headline for SEO',
      ),
    );
  });

  it('should pass persian data to the Metadata component', () => {
    doesMatch(
      MetadataWithContextAsObject('persian', articleDataPersian),
      articleMetadataBuilder(
        'persian',
        'fa',
        'سرصفحه مقاله',
        'خلاصه مقاله',
        'cwv2xv848j5o',
        'سرصفحه مقاله',
      ),
    );
  });
});
