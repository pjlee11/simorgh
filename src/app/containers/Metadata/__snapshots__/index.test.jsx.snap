// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no data should render null 1`] = `null`;
